core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -589,7 +593,10 @@ public void updateRetainedTransports(Collection<EquivalentAddressGroup> addrs) {
     @Override
     public ClientTransport getTransport(final EquivalentAddressGroup addressGroup) {
       checkNotNull(addressGroup, "addressGroup");
-      TransportSet ts;
+      TransportSet ts = transports.get(addressGroup);
```

Comment:
> If the Channel is in the process of shutting down but hasn't cleared transports, this will grab a transport.

Yes.

> Later, when each transport is being shutdown, the ClientCall will try to invoke a method on the transport and get back some sort of error, safely.

There will be a race between the creation of a new stream and the transport being shutdown. In both cases the transports must already be thread-safe to handle. If creating the new stream occurs first, then there will not be an error on the stream and shutdown waits until the stream completes. If shutdown occurs first, then the stream will immediately (during creation) be called back indicating the stream is closed (with UNAVAILABLE). Both of those are pre-existing races, although in both cases they are "safe" and within our rights permitted by the public-facing API.

> Will this PR just unmask some internal locking inside of the transports?

There is internal locking within `transports`, but only for mutations, as I commented on the field. Basically, readers use volatile reads to traverse the hash table and mutators make sure not to impact any concurrent readers. This behavior is called out in ConcurrentHashMap's JavaDoc, although I've also verified it myself:

> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove).